### ISSUE TITLE
Element.ariaHasPopup: Values not recognised by screen readers

### ISSUE THREAD

**Issue by guyhickling** [mdn/content#7102]
MDN URL: https://developer.mozilla.org/en-US/docs/Web/API/Element/ariaHasPopup

The Browser Compatibility chart on this page is unfortunately wrong because, in fact, the only values recognised by screen readers and other assistive devices are the originally specified ones of "true", "false", and "menu". 

None of the new values introduced in ARIA 1.1, dialog, listbox, tree and grid, are supported or recognised by assistive technology. Many screen readers just announce "menu" instead of any of those values.

Unfortunately, developers regularly try to use aria-haspopup="dialog" because of the frequent use of popup modal dialogs, and because of a multitude of articles on the internet recommending it. To have "menu" announced instead is confusing for blind screen reader users.

To the best of my knowledge this defect applies to all mainstream screen readers except one, and there is no known likelihood of any of them taking these values on board in the foreseeable future. 

So I recommend that:
a) you add a highly visible note to this MDN page to make clear that those other values are not supported by assistive technology and therefore should not be used
b) replace all the green markers in your chart by "partial compliance only" ones.

<details>
<summary>MDN Content page report details</summary>

* Folder: `en-us/web/api/element/ariahaspopup`
* MDN URL: https://developer.mozilla.org/en-US/docs/Web/API/Element/ariaHasPopup
* GitHub URL: https://github.com/mdn/content/blob/main/files/en-us/web/api/element/ariahaspopup/index.html
* Last commit: https://github.com/mdn/content/commit/8455a83fd1ba95f5e8cb2fadebab82cd77111579
* Document last modified: 2021-06-12T21:26:13.000Z

</details>

**Comment by sideshowbarker** [comment]
alerting @stevefaulkner and @scottaohara, who I think may have some insights about what we should do here

**Comment by scottaohara** [comment]
The browser compatibility table on this page is correct because this page is not describing the support for the attribute / its various token values, but rather the implementation support for specifying this attribute using the ARIAMixin IDL interface.

Granted, the example on this page is an odd one - why would a combobox's `aria-haspopup` value be set from `true` (incorrect for the type of popup - `listbox` - this combobox should have had) to `false` (which then means the fact the combobox has a popup wouldn't be exposed - is not how this attribute should be used).   Also, the combobox markup pattern is missing necessary attributes to make it conforming

```html
<!-- MDN example 
  incorrectly has aria-haspopup=true (set to false via the JS - should be "listbox")
  missing aria-controls
  missing aria-activedescendant
-->
<input id="animal" type="text" 
  role="combobox" aria-autocomplete="list" 
  aria-expanded="false" aria-haspopup="true">

<!-- from ARIA 1.2 spec -->
<input type="text" id="tag_combo"
      role="combobox" aria-autocomplete="list"
      aria-haspopup="listbox" aria-expanded="true"
      aria-controls="popup_listbox" aria-activedescendant="selected_option">
```

It doesn't look like MDN presently has a page that describes the `aria-haspopup` attribute ([the attribute is not linked in this section](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Techniques#widget_attributes)).  If this page existed, then its table of support would be the one to review...

On the page linked here, it's possible a note / warning could be added to the Value section to indicate that individual support for these values can vary based on the element the `aria-haspopup` attribute is specified on (but I wouldn't expect more than that per the purpose of this page). I'd also only expect a note here since the topic is not as straight forward as this issue would otherwise indicate.  Yes, a `button` with `aria-haspopup` has minimal support for the different values... and this is due to the way a `button`'s platform mapping is exposed when given this attribute (screen readers expose what they're presented).  Other elements, however, have more varied (better) support for the different attribute values.

**Comment by guyhickling** [comment]
Scott,

Thank you for your reply. I realise that this page is only describing an
interface. And of, course, this not a problem of your making. It is due to
W3C laying down a new standard for this attribute (in ARIA 1.1) , but the
browsers and assistive technology between them just ignoring it. (I think
it must be the browsers at fault primarily, since screen readers and
suchlike do announce for Apple devices, which is the only manufacturer to
have recognised it.)

However, developers use MDN as a highly respected source. And in the
absence of a page on aria-haspopup, they will use this page to obtain what
they assume is an authoritative source of what they can do. Consequently
accessibility people like myself are seeing this attribute being used more
and more, and the vast majority of the world's screen readers are giving
their users duff information when it is used. They announce the buttons as
menu buttons when they are actually opening popup dialogs. Almost every
audit I do I have to tell developers to stop using it in this scenario.

It isn't valid to say it works on some elements because, in terms of
numbers on websites, buttons opening popup dialogs probably outnumber all
other potential and actual occurrences of aria-haspopup put together! It is
such a common use case.

We have, of course, been on to the browser and assistive technology
manufacturers about this for a long time - for years now. Bugs have been
filed and complaints have been made. But they have still ignored it, and
show no signs of changing in the future. So everything should be done,
especially in documentation like MDN's that show what looks like tables of
browser support, to let developers know that the browsers do not support
it, so they should not use it.

All it needs is a note or comment to that effect on pages like this one, on
the most authoritative sources, like MDN.

Regards,
Guy Hickling
Accessibility Consultant
+44 1635 860 728

**Comment by scottaohara** [comment]
@guyhickling I'm aware of all this.  You've responded to the [ARIA issue I created on this topic almost 2 years ago](https://github.com/w3c/aria/issues/1024).

In regards to your request, per your mention of adding a note - I'm assuming you did not read the note that I added to the page in question here, via the merged PR which closed this issue?  I would suggest you do that and if you have any additional requests for that, either respond with those, or make an edit so the note meets your satisfaction. 

Regarding your comment:
>It isn't valid to say it works on some elements...

But it does work on some elements, therefore it is a valid statement. As mentioned here, and again in the linked ARIA issue, support is varied depending on the element.  Further discussion should continue to happen in the ARIA issue.  A browser support table would be insufficient to appropriately indicate the variety of support variations for this attribute, as support is based on browser + element + AT combos.

**Comment by patrickhlauke** [comment]
it's the difference between "do browsers expose/support this attribute correctly" and "do AT actually know what to do with this exposed value". two different concerns, and the page here is concerned with the former, not the latter...
